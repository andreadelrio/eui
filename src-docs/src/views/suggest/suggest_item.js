import React from 'react';

import { EuiSuggestItem, EuiSpacer } from '../../../../src/components';

const shortDescription = 'This is the description';

const sampleItems = [
  {
    type: { iconType: 'kqlField', color: 'tint4' },
    label: 'Field sample',
    description: shortDescription,
  },
  {
    type: { iconType: 'kqlValue', color: 'tint0' },
    label: 'Value sample',
    description: shortDescription,
  },
  {
    type: { iconType: 'kqlSelector', color: 'tint2' },
    label: 'Conjunction sample',
    description: shortDescription,
  },
  {
    type: { iconType: 'kqlOperand', color: 'tint1' },
    label: 'Operator sample',
    description: shortDescription,
  },
  {
    type: { iconType: 'search', color: 'tint8' },
    label: 'Recent search',
  },
  {
    type: { iconType: 'save', color: 'tint3' },
    label: 'Saved search',
  },
];

const sampleItem1 = {
  type: { iconType: 'kqlValue', color: 'tint0' },
  label: 'Charles de Gaulle International Airport',
  description: shortDescription,
};

export default () => (
  <div>
    {sampleItems.map((item, index) => (
      <EuiSuggestItem
        type={item.type}
        key={index}
        label={item.label}
        description={item.description}
      />
    ))}
    <EuiSpacer size="m" />
    <EuiSuggestItem
      type={sampleItem1.type}
      label={sampleItem1.label}
      description="This item has a fixed width label"
    />
    <EuiSuggestItem
      type={sampleItem1.type}
      labelDisplay="expand"
      label={sampleItem1.label}
      description="This item will expand its label if needed"
    />
    <EuiSpacer size="m" />
    <EuiSuggestItem
      type={{ iconType: 'search', color: 'tint8' }}
      label="Items with no description will expand their label"
    />
  </div>
);
