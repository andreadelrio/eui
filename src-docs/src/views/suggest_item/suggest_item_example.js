import React from 'react';

import { renderToHtml } from '../../services';

import { GuideSectionTypes } from '../../components';

import { EuiCode, EuiSuggestItem } from '../../../../src/components';

import SuggestItem from './suggest_item';
const suggestItemSource = require('!!raw-loader!./suggest_item');
const suggestItemHtml = renderToHtml(SuggestItem);

import SuggestItemLongLabel from './suggest_item_long_label';
const suggestItemLongLabelSource = require('!!raw-loader!./suggest_item_long_label');
const suggestItemLongLabelHtml = renderToHtml(SuggestItemLongLabel);

export const SuggestItemExample = {
  title: 'Suggest Item',
  sections: [
    {
      title: 'SuggestItem',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: suggestItemSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: suggestItemHtml,
        },
      ],
      text: (
        <div>
          <p>
            <EuiCode>EuiSuggestItem</EuiCode> is a list item component useful to
            suggestions when typing queries in input fields.
          </p>
          <ul>
            <li>
              <EuiCode>type</EuiCode> is an object that takes{' '}
              <EuiCode>icon</EuiCode> and <EuiCode>color</EuiCode>.{' '}
              <EuiCode>icon</EuiCode> supports any of the icons available in
              <EuiCode>EuiIcon</EuiCode>. <EuiCode>color</EuiCode> is the color
              for the icon and its background. The background color is
              automatically calculated and it is a ligther shade of
              <EuiCode>color</EuiCode>.
            </li>
            <li>
              <EuiCode>label</EuiCode> is the primary text for the suggestion.
            </li>
            <li>
              <EuiCode>description</EuiCode> is the secondary text for the
              suggestion and it is optional.
            </li>
            <li>
              Use <EuiCode>layout</EuiCode> to change the distribution of the{' '}
              <EuiCode>EuiSuggestItem</EuiCode> elements. The default{' '}
              <EuiCode>layout</EuiCode> is <EuiCode>setColumns</EuiCode>.
            </li>
          </ul>
        </div>
      ),
      props: { EuiSuggestItem },
      demo: <SuggestItem />,
    },
    {
      title: 'Expand long label',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: suggestItemLongLabelSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: suggestItemLongLabelHtml,
        },
      ],
      text: (
        <p>
          By default <EuiCode>EuiSuggestItem</EuiCode>&apos;s{' '}
          <EuiCode>label</EuiCode> will have a fixed width and use ellipsis
          whenever its content is too long. It is possible to show the full text
          by setting <EuiCode>layout</EuiCode> to <EuiCode>inline</EuiCode>.
        </p>
      ),
      props: { EuiSuggestItem },
      demo: <SuggestItemLongLabel />,
    },
  ],
};
