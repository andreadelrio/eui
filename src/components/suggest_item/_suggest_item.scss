@import '../loading/mixins';

// $buttonTypes: (
//   primary: $euiColorPrimary,
//   secondary: $euiColorSecondary,
//   warning: $euiColorWarning,
//   danger: $euiColorDanger,
//   dark: $euiColorMediumShade,
// );

.euiSuggestItem {
  display: flex;
  flex-grow: 1;
  align-items: center;
  font-size: $euiFontSizeXS;
  white-space: nowrap;

  @each $name, $color in $itemColors {
    .euiSuggestItem__type--#{$name} {
      background-color: tintOrShade($color, 90%, 50%);
		}
	}

  .euiSuggestItem__label,
  .euiSuggestItem__type,
  .euiSuggestItem__description {
    flex-grow: 0;
    display: flex;
    flex-direction: column;
	}

  .euiSuggestItem__type {
    position: relative;
    flex-shrink: 0;
    flex-basis: auto;
    width: $euiSizeXL;
    height: $euiSizeXL;
    text-align: center;
    overflow: hidden;
    padding: $euiSizeXS;
    justify-content: center;
    align-items: center;

    .customBackground {
      width: 32px;
      height: 32px;
      position: absolute;
      opacity: .2;
      top: 0;
      left: 0;
    }
  }

  .euiSuggestItem__label {
    flex-basis: 30%;
    font-family: $euiCodeFontFamily;
    overflow: hidden;
    text-overflow: ellipsis;
    padding: $euiSizeXS $euiSizeS;
    color: $euiTextColor;

    &.euiSuggestItem__layout--inline {
      flex-basis: auto;
    }
  }

  .euiSuggestItem__description,
  .euiSuggestItem__label {
    overflow: hidden;
    text-overflow: ellipsis;
    display: block;
	}

  .euiSuggestItem__description {
    color: $euiColorDarkShade;
    flex-basis: auto;
    padding-top: $euiSizeXS * .5;
    &:empty {
      flex-grow: 0;
      margin-left:0;
    }
	}
}
