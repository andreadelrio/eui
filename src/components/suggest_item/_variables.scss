$itemColors: (
  itemTint01: $euiColorPrimary,
  itemTint02: $euiColorSecondary,
  itemTint03: $euiColorWarning,
  itemTint04: $euiColorDanger,
  itemTint05: $euiColorMediumShade,
  itemTint06: $euiColorAccent,
  itemTint07: $euiColorVis3,
);
