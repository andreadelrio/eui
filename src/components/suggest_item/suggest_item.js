/* eslint-disable @typescript-eslint/no-unused-vars */
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { EuiIcon, IconPropType, IconColor, IconType } from '../icon';
import { TypePredicateKind } from 'typescript';

const colorToClassNameMap = {
  primary: 'euiSuggestItem__type--primary',
  secondary: 'euiSuggestItem__type--secondary',
  warning: 'euiSuggestItem__type--warning',
  danger: 'euiSuggestItem__type--danger',
  dark: 'euiSuggestItem__type--dark',
};

export const COLORS = Object.keys(colorToClassNameMap);

const layoutToClassNameMap = {
  setColumns: 'euiSuggestItem__layout--set',
  inline: 'euiSuggestItem__layout--inline',
};

export const LAYOUTS = Object.keys(layoutToClassNameMap);

export const EuiSuggestItem = ({
  className,
  label,
  type,
  layout,
  description,
  ...rest
}) => {
  const classes = classNames('euiSuggestItem', className);

  let optionalColorClass = '';
  const layoutClass = layoutToClassNameMap[layout];
  let optionalCustomStyles = undefined;

  if (type && type.color) {
    if (COLORS.indexOf(type.color) > -1) {
      optionalColorClass = colorToClassNameMap[type.color];
    } else {
      optionalCustomStyles = { backgroundColor: type.color };
    }
  }

  return (
    <div className={classes} {...rest}>
      <span className={`euiSuggestItem__type ${optionalColorClass}`}>
        {optionalCustomStyles ? (
          <div style={optionalCustomStyles} className="customBackground" />
        ) : (
          undefined
        )}
        <EuiIcon color={type.color} type={type.icon} />
      </span>
      <span className={`euiSuggestItem__label ${layoutClass}`}>{label}</span>
      <span className="euiSuggestItem__description">{description}</span>
    </div>
  );
};

EuiSuggestItem.propTypes = {
  className: PropTypes.string,
  /**
   * Takes 'icon' for EuiIcon and 'color'. 'color' can be either our palette colors (primary, secondary, etc) or a hex value.
   */
  type: PropTypes.object,
  /**
   * Label for suggestion
   */
  label: PropTypes.string,
  /**
   * Description for suggestion
   */
  description: PropTypes.string,
  /**
   * Layout for suggest item
   */
  layout: PropTypes.oneOf(LAYOUTS),
};

EuiSuggestItem.defaultProps = {
  layout: 'setColumns',
};
