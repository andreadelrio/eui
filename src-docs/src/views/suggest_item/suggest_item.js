import React from 'react';

import { EuiSuggestItem } from '../../../../src/components/suggest_item';

const shortDescription = 'This is the description';

const sampleItems = [
  {
    type: { icon: 'kqlField', color: 'warning' },
    label: 'Field sample',
    description: shortDescription,
  },
  {
    type: { icon: 'kqlValue', color: 'secondary' },
    label: 'Value sample',
    description: shortDescription,
  },
  {
    type: { icon: 'kqlSelector', color: '#7800A6' },
    label: 'Conjunction sample',
    description: shortDescription,
  },
  {
    type: { icon: 'kqlOperand', color: 'primary' },
    label: 'Operator sample',
    description: shortDescription,
  },
  {
    type: { icon: 'search', color: 'dark' },
    label: 'Recent search sample',
    description: shortDescription,
  },
  {
    type: { icon: 'save', color: '#DD0A73' },
    label: 'Saved search sample',
    description: shortDescription,
  },
];

export default () => (
  <div>
    {sampleItems.map(item => (
      <EuiSuggestItem
        type={item.type}
        label={item.label}
        description={item.description}
      />
    ))}
  </div>
);
