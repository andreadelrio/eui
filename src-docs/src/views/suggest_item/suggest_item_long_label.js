import React from 'react';

import { EuiSuggestItem } from '../../../../src/components/suggest_item';

const shortDescription = 'This is the description';

const sampleItem = {
  type: { icon: 'kqlValue', color: 'secondary' },
  label: 'Charles de Gaulle International Airport',
  description: shortDescription,
};

export default () => (
  <div>
    <EuiSuggestItem
      type={sampleItem.type}
      label={sampleItem.label}
      description={sampleItem.description}
    />
    <EuiSuggestItem
      type={sampleItem.type}
      layout="inline"
      label={sampleItem.label}
      description={sampleItem.description}
    />
  </div>
);
