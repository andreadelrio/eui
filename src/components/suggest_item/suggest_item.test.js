import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../test/required_props';

import { EuiSuggestItem } from './suggest_item';

const TYPE = {
  icon: 'search',
  color: 'primary',
};

describe('EuiSuggestItem', () => {
  test('is rendered', () => {
    const component = render(<EuiSuggestItem {...requiredProps} type={TYPE} />);

    expect(component).toMatchSnapshot();
  });
});
