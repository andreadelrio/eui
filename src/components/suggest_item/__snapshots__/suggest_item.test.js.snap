// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiSuggestItem is rendered 1`] = `
<div
  aria-label="aria-label"
  class="euiSuggestItem testClass1 testClass2"
  data-test-subj="test subject string"
>
  <span
    class="euiSuggestItem__type euiSuggestItem__type--primary"
  >
    <svg
      class="euiIcon euiIcon--medium euiIcon--primary euiIcon-isLoading"
      focusable="false"
      height="16"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    />
  </span>
  <span
    class="euiSuggestItem__label euiSuggestItem__layout--set"
  />
  <span
    class="euiSuggestItem__description"
  />
</div>
`;
